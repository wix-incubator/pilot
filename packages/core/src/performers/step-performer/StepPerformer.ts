import { StepPerformerPromptCreator } from "./StepPerformerPromptCreator";
import { CodeEvaluator } from "@/common/CodeEvaluator";
import { CacheHandler } from "@/common/cacheHandler/CacheHandler";
import { SnapshotComparator } from "@/common/snapshot/comparator/SnapshotComparator";
import {
  CodeEvaluationResult,
  PreviousStep,
  PromptHandler,
  ScreenCapturerResult,
  StepPerformerCacheValue,
} from "@/types";
import { extractPilotOutputs } from "@/common/extract/extractTaggedOutputs";
import { ScreenCapturer } from "@/common/snapshot/ScreenCapturer";
import logger from "@/common/logger";

export class StepPerformer {
  private sharedContext: Record<string, any> = {};

  constructor(
    private context: any,
    private promptCreator: StepPerformerPromptCreator,
    private codeEvaluator: CodeEvaluator,
    private promptHandler: PromptHandler,
    private cacheHandler: CacheHandler,
    private snapshotComparator: SnapshotComparator,
    private screenCapturer: ScreenCapturer,
  ) {}

  extendJSContext(newContext: any): void {
    for (const key in newContext) {
      if (key in this.context) {
        logger
          .labeled("WARNING")
          .warn(
            `Pilot's variable from context \`${key}\` is overridden by a new value from \`extendJSContext\``,
          );
        break;
      }
    }
    this.context = { ...this.context, ...newContext };
  }

  private async generateCode(
    currentStep: string,
    previousSteps: PreviousStep[],
    screenCapture: ScreenCapturerResult,
  ): Promise<any> {
    const cacheKey = this.cacheHandler.generateCacheKey({
      currentStep,
      previousSteps,
    });

    if (this.cacheHandler.isCacheInUse() && cacheKey) {
        logger.warn(
            `I AM LOOKING FOR MATCHING CACHE VALUE ${JSON.stringify(cacheKey)}`,
        );
      const cachedValues =
        this.cacheHandler.getFromPersistentCache<StepPerformerCacheValue>(
          cacheKey,
        );
      if (cachedValues) {
          logger.warn(
              `I AM LOOKING FOR MATCHING ENTRY ${JSON.stringify(cacheKey)}`,
          );
        const matchingEntry =
          await this.cacheHandler.findMatchingCacheEntryValidationMatcherBased<StepPerformerCacheValue>(
            cachedValues,
            this.context,
            this.sharedContext,
          );

        if (matchingEntry) {
          logger.warn(`I WAS TAKEN FROM CACHE ${matchingEntry.value.code}`);
          return { code: matchingEntry.value.code, shouldRunMoreCode: matchingEntry.shouldRunMoreCode };
        }
      }
    }

    // No cache match found, generate new code
    const prompt = this.promptCreator.createPrompt(
      currentStep,
      screenCapture.viewHierarchy ?? "Unknown view hierarchy",
      !!screenCapture.snapshot,
      previousSteps,
    );

    const promptResult = await this.promptHandler.runPrompt(
      prompt,
      screenCapture.snapshot,
    );

    const extractedCodeBlock = extractPilotOutputs(promptResult);

    const code = extractedCodeBlock.code
      ? extractedCodeBlock.code
      : "No code found";
    const cacheValidationMatcher = extractedCodeBlock.cacheValidationMatcher
      ? extractedCodeBlock.cacheValidationMatcher
      : ["Unknown view hierarchy"];

    if (this.cacheHandler.isCacheInUse() && cacheKey) {
      const cacheValue: StepPerformerCacheValue = { code };
      this.cacheHandler.addToTemporaryCacheValidationMatcherBased(
        cacheKey,
        cacheValue,
        cacheValidationMatcher,
      );
    }
    return { code: code, cacheValidationMatcher: cacheValidationMatcher, shouldRunMoreCode: true };
  }

  async perform(
    step: string,
    previous: PreviousStep[] = [],
    screenCapture: ScreenCapturerResult,
    maxAttempts: number = 2,
  ): Promise<CodeEvaluationResult> {
    const progress = logger.startProgress(
      {
        actionLabel: "STEP",
        successLabel: "DONE",
        failureLabel: "FAIL",
      },
      {
        message: step,
        isBold: true,
        color: "whiteBright",
      },
    );
    let lastError: any = null;
    let lastCode: string | undefined;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        const screenCaptureResult =
          attempt == 1
            ? screenCapture
            : await this.screenCapturer.capture(true);

        const generatedCode = await this.generateCode(
          step,
          previous,
          screenCaptureResult,
        );

        const code= generatedCode.code;
        const shouldRunMoreCode = generatedCode.shouldRunMoreCode;
        lastCode = code;

        if (!code) {
          progress.updateLabel("RETRY", {
            message: step,
            isBold: true,
            color: "yellow",
          });

          throw new Error(
            "Failed to generate code from intent, please retry generating the code or provide a code that throws a descriptive error.",
          );
        }

          let result: CodeEvaluationResult;
          if (!shouldRunMoreCode) {
              result = {
                  code,
                  result: undefined, // or some placeholder
                  sharedContext: this.sharedContext,
              };
          } else {
              result = await this.codeEvaluator.evaluate(
                  code,
                  this.context,
                  this.sharedContext,
              );
              this.sharedContext = result.sharedContext || this.sharedContext;
          }
        progress.stop("success", {
          message: "Step completed successfully",
          isBold: true,
          color: "green",
        });

        if (attempt > 1) {
          logger
            .labeled("SUCCESS")
            .info(
              `Attempt ${attempt}/${maxAttempts} succeeded for step "${step}"`,
            );
        }

        return result;
      } catch (error) {
        lastError = error;
        const errorDetails = error instanceof Error ? error.message : error;
        logger
          .labeled("ERROR")
          .error(
            `Attempt ${attempt}/${maxAttempts} failed for step: ${step}, with error: ${errorDetails}`,
          );

        if (attempt < maxAttempts) {
          progress.updateLabel("RETRY", "Trying again");

          previous = [
            ...previous,
            {
              step,
              code: lastCode ?? "undefined",
              error: errorDetails,
            },
          ];
        }
      }
    }

    progress.stop("failure", "Step failed after multiple attempts");
    throw lastError;
  }
}
