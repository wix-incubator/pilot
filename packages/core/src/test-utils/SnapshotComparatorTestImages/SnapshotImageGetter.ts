import path from "path";

type image =
  | "baseline"
  | "different"
  | "grayed"
  | "very_different_colors"
  | "with_text";
const imageFileNames: Record<image, string> = {
  baseline: "baseline.png",
  different: "different.png",
  grayed: "grayed.png",
  very_different_colors: "very_different_colors.png",
  with_text: "with_text.png",
};

export const getSnapshotImage = (image: image) => {
  return path.resolve(__dirname, ".", imageFileNames[image]);
};
