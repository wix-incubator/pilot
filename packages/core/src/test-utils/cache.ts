import fs from "fs";
import * as fileSystem from "../common/cacheHandler/fileSystem";

export let mockedCacheFile: { [key: string]: any } | undefined;

export const mockCache = (
  data: { [key: string]: any } | undefined = undefined,
) => {
  mockedCacheFile = data;

  // Mock fs methods
  (fs.writeFileSync as jest.Mock).mockImplementation((filePath, data) => {
    mockedCacheFile = JSON.parse(data);
  });

  (fs.readFileSync as jest.Mock).mockReturnValue(
    JSON.stringify(mockedCacheFile),
  );

  (fs.existsSync as jest.Mock).mockReturnValue(mockedCacheFile !== undefined);
  
  // Mock fileSystem module methods
  jest.spyOn(fileSystem, 'writeJsonFile').mockImplementation((filePath, data) => {
    mockedCacheFile = data as { [key: string]: any };
    return true;
  });
  
  jest.spyOn(fileSystem, 'readJsonFile').mockImplementation(() => {
    return mockedCacheFile;
  });
  
  jest.spyOn(fileSystem, 'ensureDirectoryExists').mockImplementation(() => {
    // Do nothing
  });
};
