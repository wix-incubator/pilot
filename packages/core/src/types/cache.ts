import { AutoReview, AutoStepPlan } from "@/types/auto";

/**
 * Screen capture output.
 */
export type ScreenCapturerResult = {
  /** UI snapshot path */
  snapshot: string | undefined;
  /** Component hierarchy */
  viewHierarchy: string;
  /** Image capture support status */
  isSnapshotImageAttached: boolean;
};

/** Snapshot hashing algorithm types */
export type HashingAlgorithm = "BlockHash";

/** Hash algorithm output format */
export type SnapshotHashObject = Record<HashingAlgorithm, string>;

/**
 * Shared snapshot data used in both SingleCacheValue and SingleAutoPilotCacheValue
 */
export type SnapshotData = {
  /** UI snapshot hash */
  snapshotHash?: SnapshotHashObject;
  /** View hierarchy hash (MD5) */
  viewHierarchyHash: string;
};

/**
 * Single cache entry for step performer.
 */
export type SingleCacheValue = SnapshotData & {
  /** Generated code */
  code: string;
};

/**
 * Auto pilot cache entry with minimal required data
 */
export type SingleAutoPilotCacheValue = SnapshotData & {
  /** Screen description */
  screenDescription: string;
  /** Auto pilot step plan */
  plan: AutoStepPlan;
  /** Auto pilot step review */
  review: AutoReview;
  /** Goal achievement status */
  goalAchieved: boolean;
  /** Success summary */
  summary?: string;
};

/** Cache entry array */
export type CacheValues = SingleCacheValue[];

/** Auto pilot cache entry array */
export type CacheAutoPilotValues = SingleAutoPilotCacheValue[];

/**
 * Snapshot hashing operations interface.
 */
export interface SnapshotHashing {
  /**
   * Generates hash from UI snapshot.
   * @param snapshot - UI snapshot to hash
   * @returns Hash string
   */
  hashSnapshot(snapshot: any): Promise<string>;

  /**
   * Calculates difference between snapshot hashes.
   * @param hash1 - First snapshot hash
   * @param hash2 - Second snapshot hash
   * @returns Numerical difference
   */
  calculateSnapshotDistance(hash1: string, hash2: string): number;

  /**
   * Checks if snapshots are similar enough.
   * @param hash1 - First snapshot hash
   * @param hash2 - Second snapshot hash
   * @param threshold - Optional similarity threshold
   * @returns true if similar, false otherwise
   */
  areSnapshotsSimilar(
    hash1: string,
    hash2: string,
    threshold?: number,
  ): boolean;
}
