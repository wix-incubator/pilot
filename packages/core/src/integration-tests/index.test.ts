import { Pilot } from "@/Pilot";
import fs from "fs";
import {
  PromptHandler,
  TestingFrameworkDriver,
  AutoReport,
  CacheValue,
} from "@/types";
import { mockedCacheFile, mockCache } from "@/test-utils/cache";
import { StepPerformerPromptCreator } from "@/performers/step-performer/StepPerformerPromptCreator";
import { StepPerformer } from "@/performers/step-performer/StepPerformer";
import {
  bazCategory,
  barCategory1,
  dummyContext,
} from "@/test-utils/APICatalogTestUtils";
import { getSnapshotImage } from "@/test-utils/SnapshotComparatorTestImages/SnapshotImageGetter";

jest.mock("crypto");
jest.mock("fs");

const CACHE_VALIDATION_TAG =
  "<CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER>";
const CACHE_VALIDATION_CODE = "verify button exists";

describe("Pilot Integration Tests", () => {
  let mockFrameworkDriver: jest.Mocked<TestingFrameworkDriver>;
  let mockPromptHandler: jest.Mocked<PromptHandler>;
  let pilot: Pilot;

  beforeEach(async () => {
    jest.clearAllMocks();

    mockFrameworkDriver = {
      captureSnapshotImage: jest
        .fn()
        .mockResolvedValue(getSnapshotImage("baseline")),
      captureViewHierarchyString: jest
        .fn()
        .mockResolvedValue("<view><button>Login</button></view>"),
      apiCatalog: {
        context: {},
        categories: [],
      },
      driverConfig: { useSnapshotStabilitySync: true },
    };

    mockPromptHandler = {
      runPrompt: jest.fn(),
      isSnapshotImageSupported: jest.fn().mockReturnValue(true),
    };

    mockCache();

    pilot = new Pilot({
      frameworkDriver: mockFrameworkDriver,
      promptHandler: mockPromptHandler,
    });
  });

  describe("Basic Operations", () => {
    it("should be properly instantiated", () => {
      expect(pilot).toBeInstanceOf(Pilot);
    });
  });

  describe("Single Step Operations", () => {
    beforeEach(() => {
      pilot.start();
    });

    it("should successfully perform an action", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// No operation</CODE>${CACHE_VALIDATION_TAG}`,
      );
      await expect(
        pilot.perform("Tap on the login button"),
      ).resolves.not.toThrow();

      expect(mockFrameworkDriver.captureSnapshotImage).toHaveBeenCalled();
      expect(mockFrameworkDriver.captureViewHierarchyString).toHaveBeenCalled();
      expect(mockPromptHandler.runPrompt).toHaveBeenCalledWith(
        expect.stringContaining("Tap on the login button"),
        getSnapshotImage("baseline"),
      );
    });

    it("should successfully perform an assertion", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// No operation</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await expect(
        pilot.perform("The welcome message should be visible"),
      ).resolves.not.toThrow();

      expect(mockFrameworkDriver.captureSnapshotImage).toHaveBeenCalled();
      expect(mockFrameworkDriver.captureViewHierarchyString).toHaveBeenCalled();
      expect(mockPromptHandler.runPrompt).toHaveBeenCalledWith(
        expect.stringContaining("The welcome message should be visible"),
        getSnapshotImage("baseline"),
      );
    });

    it("should handle errors during action execution", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>throw new Error("Element not found")</CODE>${CACHE_VALIDATION_TAG}`,
      );
      await expect(
        pilot.perform("Tap on a non-existent button"),
      ).rejects.toThrow("Element not found");
    });

    it("should handle errors during assertion execution", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>throw new Error("Element not found")</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await expect(
        pilot.perform("The welcome message should be visible"),
      ).rejects.toThrow("Element not found");
    });

    it("should handle errors during code evaluation", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>foobar</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await expect(
        pilot.perform("The welcome message should be visible"),
      ).rejects.toThrow(/foobar is not defined/);
    });
  });

  describe("Multiple Step Operations", () => {
    beforeEach(() => {
      pilot.start();
    });

    it("should perform multiple steps using spread operator", async () => {
      mockPromptHandler.runPrompt
        .mockResolvedValueOnce(
          `<CODE>// Tap login button</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>// Enter username</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>// Enter password</CODE>${CACHE_VALIDATION_TAG}`,
        );

      await pilot.perform(
        "Tap on the login button",
        'Enter username "testuser"',
        'Enter password "password123"',
      );

      expect(mockPromptHandler.runPrompt).toHaveBeenCalledTimes(3);
      expect(mockFrameworkDriver.captureSnapshotImage).toHaveBeenCalledTimes(6);
      expect(
        mockFrameworkDriver.captureViewHierarchyString,
      ).toHaveBeenCalledTimes(3);
    });

    it("should handle errors in multiple steps and stop execution", async () => {
      mockPromptHandler.runPrompt
        .mockResolvedValueOnce(
          `<CODE>// Tap login button</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>throw new Error("Username field not found");</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>throw new Error("Username field not found - second");</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>// Enter password</CODE>${CACHE_VALIDATION_TAG}`,
        );

      await expect(
        pilot.perform(
          "Tap on the login button",
          'Enter username "testuser"',
          'Enter password "password123"',
        ),
      ).rejects.toThrow("Username field not found");

      expect(mockPromptHandler.runPrompt).toHaveBeenCalledTimes(3);
      expect(mockFrameworkDriver.captureSnapshotImage).toHaveBeenCalledTimes(6);
      expect(
        mockFrameworkDriver.captureViewHierarchyString,
      ).toHaveBeenCalledTimes(3);
    });
  });

  describe("Error Handling", () => {
    beforeEach(() => {
      pilot.start();
    });

    it("should throw error when PromptHandler fails", async () => {
      mockPromptHandler.runPrompt.mockRejectedValue(new Error("API error"));

      await expect(pilot.perform("Perform action")).rejects.toThrow(
        "API error",
      );
    });

    it("should throw error when captureSnapshotImage() fails", async () => {
      mockFrameworkDriver.captureSnapshotImage.mockRejectedValue(
        new Error("Snapshot error"),
      );

      await expect(pilot.perform("Perform action")).rejects.toThrow(
        "Snapshot error",
      );
    });

    it("should throw error when captureViewHierarchyString() fails", async () => {
      mockFrameworkDriver.captureViewHierarchyString.mockRejectedValue(
        new Error("Hierarchy error"),
      );

      await expect(pilot.perform("Perform action")).rejects.toThrow(
        "Hierarchy error",
      );
    });
  });

  describe("Context Management", () => {
    beforeEach(() => {
      pilot.start();
    });

    it("should reset context when end is called", async () => {
      mockPromptHandler.runPrompt.mockResolvedValueOnce(
        `<CODE>// Login action</CODE>${CACHE_VALIDATION_TAG}`,
      );
      await pilot.perform("Log in to the application");

      pilot.end();
      pilot.start();

      mockPromptHandler.runPrompt.mockResolvedValueOnce(
        `<CODE>// New action after reset </CODE>${CACHE_VALIDATION_TAG}`,
      );
      await pilot.perform("Perform action after reset");

      expect(mockPromptHandler.runPrompt).toHaveBeenCalledTimes(2);
      expect(mockPromptHandler.runPrompt.mock.calls[1][0]).not.toContain(
        "Log in to the application",
      );
    });

    it("should clear conversation history on reset", async () => {
      mockPromptHandler.runPrompt
        .mockResolvedValueOnce(
          `<CODE>// Action 1</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockResolvedValueOnce(
          `<CODE>// Action 2</CODE>${CACHE_VALIDATION_TAG}`,
        );

      await pilot.perform("Action 1");
      await pilot.perform("Action 2");

      const lastCallArgsBeforeReset =
        mockPromptHandler.runPrompt.mock.calls[1][0];
      expect(lastCallArgsBeforeReset).toContain("Action 1");
      expect(lastCallArgsBeforeReset).toContain("Action 2");

      pilot.end();
      pilot.start();

      mockPromptHandler.runPrompt.mockResolvedValueOnce(
        `<CODE>// New action</CODE>${CACHE_VALIDATION_TAG}`,
      );
      await pilot.perform("New action after reset");

      const lastCallArgsAfterReset =
        mockPromptHandler.runPrompt.mock.calls[2][0];
      expect(lastCallArgsAfterReset).not.toContain("Action 1");
      expect(lastCallArgsAfterReset).not.toContain("Action 2");
      expect(lastCallArgsAfterReset).toContain("New action after reset");
    });
  });

  describe("Caching Behavior", () => {
    beforeEach(() => {
      // Create a new pilot instance with default cache options
      pilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
      });
      pilot.start();
    });

    it("should create cache file if it does not exist", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// Perform action</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await pilot.perform("Perform action");
      pilot.end(true);

      expect(mockedCacheFile).toEqual({
        '{"currentStep":"Perform action","previousSteps":[]}':
          expect.arrayContaining([
            expect.objectContaining({
              value: {
                code: "// Perform action",
              },
              validationMatcher: "No cache validation matcher found",
            }),
          ]),
      });
    });

    it("should use snapshot cache if available", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// New action code</CODE><CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER>`,
      );
      mockCache({
        '{"currentStep":"Cached action","previousSteps":[]}': [
          {
            value: {
              code: "// Cached action code",
            },
            validationMatcher: CACHE_VALIDATION_CODE,
            creationTime: Date.now(),
          },
        ],
      });

      // restart pilot to reload cache file
      pilot.end();
      pilot.start();

      await pilot.perform("Cached action");

      expect(mockPromptHandler.runPrompt).toHaveBeenCalledTimes(1);
    });

    it("should update cache file after performing new action", async () => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// New action code</CODE><CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER>`,
      );

      await pilot.perform("New action");
      pilot.end();

      expect(mockedCacheFile).toEqual({
        '{"currentStep":"New action","previousSteps":[]}': [
          expect.objectContaining({
            value: {
              code: "// New action code",
            },
            validationMatcher: "No cache validation matcher found",
          }),
        ],
      });
    });

    it("should handle fs.readFileSync errors", async () => {
      mockCache({}); // Set up an initial mocked file
      (fs.readFileSync as jest.Mock).mockImplementation(() => {
        throw new Error("Read error");
      });
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// New action code</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await pilot.perform("Action with read error");

      expect(mockPromptHandler.runPrompt).toHaveBeenCalled();
    });

    it("should handle fs.writeFileSync errors", async () => {
      mockCache(undefined); // No mocked file exists
      (fs.writeFileSync as jest.Mock).mockImplementation(() => {
        throw new Error("Write error");
      });
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// Action code</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await expect(
        pilot.perform("Action with write error"),
      ).resolves.not.toThrow();
    });

    it("should add error and not add result for next perform step", async () => {
      let promptParam: string = "";
      mockPromptHandler.runPrompt
        .mockResolvedValueOnce(
          `<CODE>throw new Error("Element not found");</CODE>${CACHE_VALIDATION_TAG}`,
        )
        .mockImplementationOnce((prompt, _snapshot) => {
          promptParam = prompt;
          return Promise.resolve(
            `<CODE>// No operation</CODE>${CACHE_VALIDATION_TAG}`,
          );
        });

      await pilot.perform("Tap on a non-existent button");

      expect(promptParam).toContain("Element not found");
      expect(promptParam).toMatchSnapshot();
    });
  });

  describe("Feature Support", () => {
    beforeEach(() => {
      // Create a new instance for each test
      pilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
      });
      pilot.start();
    });

    it("should work without snapshot images when not supported", async () => {
      mockPromptHandler.isSnapshotImageSupported.mockReturnValue(false);
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// Perform action without snapshot</CODE>${CACHE_VALIDATION_TAG}`,
      );

      await pilot.perform("Perform action without snapshot support");

      expect(mockFrameworkDriver.captureSnapshotImage).not.toHaveBeenCalled();
      expect(mockFrameworkDriver.captureViewHierarchyString).toHaveBeenCalled();
      expect(mockPromptHandler.runPrompt).toHaveBeenCalledWith(
        expect.stringContaining("Perform action without snapshot support"),
        undefined,
      );
    });
  });

  describe("API Catalog Extension", () => {
    beforeEach(() => {
      jest.clearAllMocks();
      // Create new pilot instance
      pilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
      });
      pilot.start();
    });

    it("should call relevant functions to extend the catalog", () => {
      const spyPromptCreator = jest.spyOn(
        StepPerformerPromptCreator.prototype,
        "extendAPICategories",
      );
      const spyStepPerformer = jest.spyOn(
        StepPerformer.prototype,
        "extendJSContext",
      );

      pilot.extendAPICatalog([bazCategory]);
      expect(spyPromptCreator).toHaveBeenCalledTimes(1);

      pilot.extendAPICatalog([barCategory1], dummyContext);
      expect(spyPromptCreator).toHaveBeenCalledTimes(2);
      expect(spyStepPerformer).toHaveBeenCalledTimes(1);
    });
  });

  describe("Autopilot Method", () => {
    beforeEach(() => {
      jest.clearAllMocks();

      mockPromptHandler = {
        runPrompt: jest.fn(),
        isSnapshotImageSupported: jest.fn(),
      } as any;

      mockFrameworkDriver = {
        apiCatalog: {
          context: {},
          categories: [],
        },
        captureSnapshotImage: jest.fn(),
        captureViewHierarchyString: jest.fn(),
        driverConfig: { useSnapshotStabilitySync: true },
      };

      // Use the standard pilot instance
      pilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
      });
      pilot.start();
    });

    it("should perform autopilot flow and return a pilot report", async () => {
      const goal = "Complete the login flow";
      const mockPilotReport: AutoReport = {
        summary: "All steps completed successfully",
        goal: goal,
        steps: [
          {
            screenDescription: "Login Screen",
            plan: {
              thoughts: "First step thoughts",
              action: "Tap on login button",
            },
            code: "First step code output",
            review: {
              ux: {
                summary: "UX review for first step",
                findings: [],
                score: "7/10",
              },
              a11y: {
                summary: "Accessibility review for first step",
                findings: [],
                score: "8/10",
              },
            },
            goalAchieved: true,
          },
        ],
        review: {
          ux: {
            summary: "Overall UX review",
            findings: [],
            score: "9/10",
          },
          a11y: {
            summary: "Overall Accessibility review",
            findings: [],
            score: "9/10",
          },
        },
      };

      const spyPilotPerformerPerform = jest
        .spyOn(pilot["autoPerformer"], "perform")
        .mockResolvedValue(mockPilotReport);

      const result = await pilot.autopilot(goal);

      expect(spyPilotPerformerPerform).toHaveBeenCalledTimes(1);
      expect(spyPilotPerformerPerform).toHaveBeenCalledWith(goal, undefined);
      expect(result).toEqual(mockPilotReport);
    });

    it("should handle errors from autoPerformer.perform", async () => {
      const goal = "Some goal that causes an error";
      const errorMessage = "Error during autopilot execution";

      const spyPilotPerformerPerform = jest
        .spyOn(pilot["autoPerformer"], "perform")
        .mockRejectedValue(new Error(errorMessage));

      await expect(pilot.autopilot(goal)).rejects.toThrow(errorMessage);

      expect(spyPilotPerformerPerform).toHaveBeenCalledTimes(1);
      expect(spyPilotPerformerPerform).toHaveBeenCalledWith(goal, undefined);
    });
  });

  describe("Cache Modes", () => {
    beforeEach(() => {
      mockPromptHandler.runPrompt.mockResolvedValue(
        `<CODE>// No operation</CODE>${CACHE_VALIDATION_TAG}`,
      );
    });

    it("should use cache by default", async () => {
      // Create a new instance with default cache options
      const cachePilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
      });
      cachePilot.start();

      await cachePilot.perform("Tap on the login button");
      cachePilot.end();

      const firstCacheValue = Object.values(
        (mockedCacheFile as Record<string, CacheValue<any>[]>) || {},
      )[0][0];

      expect(firstCacheValue).toHaveProperty("validationMatcher");
      expect(firstCacheValue).toHaveProperty("value");
    });

    it("should not use cache when cache mode is disabled", async () => {
      // Create a new instance with cache disabled
      const noCachePilot = new Pilot({
        frameworkDriver: mockFrameworkDriver,
        promptHandler: mockPromptHandler,
        options: {
          cacheOptions: { shouldUseCache: false },
        },
      });
      noCachePilot.start();

      // First call
      await noCachePilot.perform("Tap on the login button");
      noCachePilot.end();

      // Second call with same intent
      noCachePilot.start();
      await noCachePilot.perform("Tap on the login button");
      noCachePilot.end();

      // Should call runPrompt twice since cache is disabled
      expect(mockPromptHandler.runPrompt).toHaveBeenCalledTimes(2);
    });
  });
});
