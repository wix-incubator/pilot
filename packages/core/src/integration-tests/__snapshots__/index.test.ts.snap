// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pilot Integration Tests Caching Behavior should add error and not add result for next perform step 1`] = `
"# Test Code Generation

You are an AI assistant tasked with generating test code for a specific test step using the mentioned testing framework.
Generate the minimal executable code to perform the desired intent, based on the provided step and context.

## Context

### Intent to perform

Generate the minimal executable code to perform the following intent: "Tap on a non-existent button"

### View hierarchy

This is the complete view hierarchy. Use only the relevant parts for the executable code.
\`\`\`
<view><button>Login</button></view>
\`\`\`

### Snapshot image

A snapshot image is attached for visual reference.

### Previous steps

#### Step 1
- Intent: "Tap on a non-existent button"
- Generated code:
\`\`\`
throw new Error("Element not found");
\`\`\`
- Error occurred in your previous attempt. Try another approach to perform this step. Error message:
\`\`\`
Element not found
\`\`\`


## Available Testing Framework API


## Instructions

Your task is to generate the minimal executable code to perform the following intent: "Tap on a non-existent button". The code should include appropriate synchronization using the testing framework's wait methods to ensure reliable test execution. In addition, inside <CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER>, add code lines that verify the existence of the elements that will be interacted with in this step. This validation matcher will be used in subsequent test executions to wait for the elements to appear on the screen.

Please follow these steps carefully:

1. Analyze the provided intent, the view hierarchy, and the snapshot image to understand the required action.
2. When interacting with an element, ensure that you use the correct identifier from the view hierarchy. Do not rely on a screenshot to guess the element's selectors. Add code lines that verify the existence of the elements that will be interacted with in this step inside <CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER>.
3. Include appropriate synchronization in your code using the testing framework's wait methods to ensure elements are present and ready before interacting with them. Use the testing framework's documented wait APIs to make the test more reliable and prevent flaky failures.
4. The code inside <CACHE_VALIDATION_MATCHER></CACHE_VALIDATION_MATCHER> will be cached and used in future test runs to quickly verify that the page is in the expected state before executing the main test logic.
5. Assess the positions of elements within the screen layout. Ensure that tests accurately reflect their intended locations, such as whether an element is centered or positioned relative to others. Tests should fail if the actual locations do not align with the expected configuration.
6. Determine if the intent can be fully validated visually using the snapshot image.
7. If the intent can be visually analyzed and passes the visual check, return only comments explaining the successful visual assertion.
8. If the visual assertion fails, return code that throws an informative error explaining the failure, inside <CODE></CODE> block.
9. If visual validation is not possible, proceed to generate the minimal executable code required to perform the intent.
10. If you cannot generate the relevant code due to ambiguity or invalid intent, return code that throws an informative error explaining the problem in one sentence.
11. Each step must be completely independent - do not rely on any variables or assignments from previous steps. Even if a variable was declared or assigned in a previous step, you must redeclare and reassign it in your current step.
12. Use the provided the testing framework APIs as much as possible - prefer using the documented API methods over creating custom implementations.
13. If you need to share data between steps, use the 'sharedContext' object. You can access and modify it directly like: sharedContext.myKey = 'myValue'
14. Wrap the generated code with <CODE></CODE> block, without any additional formatting.
15. Do not provide any additional code beyond the minimal executable code required to perform the intent.

### Verify the prompt

Before generating the code, please review the provided context and instructions to ensure they are clear and unambiguous. If you encounter any issues or have questions, please throw an informative error explaining the problem.

### Examples

#### Example of throwing an informative error:
<CODE>
throw new Error("Unable to find the 'Submit' button element in the current context.");
</CODE>

#### Example of providing the validation matcher
<CACHE_VALIDATION_MATCHER>
const page = getCurrentPage(); const inputElement = await findElement(page, {placeholder: "Type the domain you want","aria-label": "Type the domain you want",class: "KvoMHf has-custom-focus wixui-text-input__input"}) ?? (() => { throw new Error('Input not found'); })();
</CACHE_VALIDATION_MATCHER>
#### Visual validation using the snapshot:
\`\`\`typescript
// Visual assertion passed based on the snapshot image.
\`\`\`

Please provide your code below:"
`;
